master/static/srv/shell-entrypoint.sh

```diff
@@ -44,8 +44,19 @@ test -f "${STARTUP_HOOK}" && source "${STARTUP_HOOK}"
 # After openssh 8+ is the only version of openssh supported (that is, after we
 # only support ubuntu >= 20.04), we can use the more obvious SetEnv option and
 # skip this awkwardness.
-blacklist="^(_|HOME|TERM|LANG|LC_.*)"
-vars="$(env | sed -E -e "s/=.*//; /$blacklist/d")"
+#
+# For HPC systems, bash module support uses variables that store functions
+# of the form below (with embedded parenthesis or %% in the name).
+#   BASH_FUNC_ml()=() {  eval $($LMOD_DIR/ml_cmd "$@")
+#   BASH_FUNC_module%%=() {  eval `/opt/lib/modulecmd bash $*`
+# so we add variables with parens or % in the name to the blacklist and 
+# filter them out because the additional parens/% in the variable names 
+# causes problems in the escaping code below.   Additionally, we properly
+# enumerate just the names by adding the grep '^\S+=\S' below to remove 
+# the multi-line values from such functions so the blacklist can be
+# properly applied.
+blacklist="^(_|HOME|TERM|LANG|\S+\(\).*|\S+\%\S*|LC_.*)"
+vars="$(env | grep -E '^\S+=\S' | sed -E -e "s/=.*//; /$blacklist/d")"
```

Comment:
I'm thinking that we've hit the limit of this blacklist/sed strategy for parsing environment variables.  For example, we are acknowledging that we are dealing with multiline variables, but the grep command is assuming the multiline variables never match `^\S+=\S`.

How about we use `env -0` to get NUL-byte delimited environment variables, then use `mapfile -d ''` to read them into a bash array, then we iterate through the bash array:

```bash
# extglob enables +() notation in patterns of ${parameter/pattern/string} notation
shopt -s extglob

# convert NUL-delimited environment key-value pairs in an array
# -d '' means "use NUL byte as delimiter"
# -t means "strip the delimiter"
mapfile -d '' -t kvps < <(env -0)

# iterate through each key-value pair in the array
options="$(
    for kvp in "${kvps[@]}" ; do
        # Variable name is what comes before the first '='
        var="${kvp/=*/}"
        # Variable content starts after the first '='
        val="${kvp/#+([^=])=/}"

        # Filter names we shouldn't forward
        if [[ "$var" =~ ^(_|HOME|TERM|LANG|LC_.*)$ ]] ; then
            continue
        fi

        # For slurm: filter variables with %% or () in the name
        if [[ "$var" =~ (%%|\(\)) ]] ; then
            continue
        fi

        # Backslash-escape quotes so that sshd works.
        val="${val//\"/\\\"}"
        # Backslash-escape backslashes so that sed doesn't interpret them.
        val="${val//\\/\\\\}"
        # Backslash-escape forward slashes so that sed works.
        val="${val////\\/}"
        echo -n "environment=\"$var=$val\","
    done
)"
```

I tested with:
```
env -i \
 LC_ALL=1 \
 ASDF=A=B=C \
 "$(echo -e 'newline=before\nafter')" \
 "PERCENTS%%=ZXCV" \
 "PARENS()=ZXCV" \
 bash test.sh
```

to confirm that equals and newlines in variables are preserved, but LC_* is still ignored, as are %% and () if they appear.

Reply:
Done